py/kubeflow/kubeflow/ci/update_jupyter_web_app.py

```diff
@@ -0,0 +1,269 @@
+
+Requires python3
+
+hub CLI depends on an OAuth token with repo permissions:
+https://hub.github.com/hub.1.html
+  * It will look for environment variable GITHUB_TOKEN
+"""
+
+import logging
+import os
+import re
+import tempfile
+import yaml
+
+import fire
+import git
+import httplib2
+
+from kubeflow.testing import util # pylint: disable=no-name-in-module
+
+from containerregistry.client import docker_creds
+from containerregistry.client import docker_name
+from containerregistry.client.v2_2 import docker_http
+from containerregistry.client.v2_2 import docker_image as v2_2_image
+from containerregistry.transport import transport_pool
+
+class WebAppUpdater(object): # pylint: disable=useless-object-inheritance
+  def __init__(self):
+    self._last_commit = None
+
+  def build_image(self, build_project, registry_project):
+    """Build the image.
+
+    Args:
+      build_project: GCP project used to build the image.
+      registry_project: GCP project used to host the image.
+    """
+    env = dict()
+    env.update(os.environ)
+    env["PROJECT"] = build_project
+    env["REGISTRY_PROJECT"] = registry_project
+
+    with tempfile.NamedTemporaryFile() as hf:
+      name = hf.name
+    env["OUTPUT"] = name
+    web_dir = self._component_dir()
+    util.run(["make", "build-gcb"], env=env, cwd=web_dir)
+
+    with open(name) as hf:
+      data = yaml.load(hf)
+
+    return data["image"]
+
+  def update_prototype(self, image):
+    values = {"image": image}
+
+    regexps = {}
+    for param, value in values.items():
+      r = re.compile(r"([ \t]*" + param + ":+ ?\"?)[^\",]+(\"?,?)")
+      v = r"\g<1>" + value + r"\2"
+      regexps[param] = (r, v, value)
+
+    prototype_file = os.path.join(self._root_dir(),
+                                  "kubeflow/jupyter/prototypes",
+                                  "jupyter-web-app.jsonnet")
+    with open(prototype_file) as f:
+      prototype = f.read().split("\n")
+    replacements = 0
+    for i, line in enumerate(prototype):
+      for param, _ in regexps.items():
+        if param not in line:
+          continue
+        if line.startswith("//"):
+          prototype[i] = re.sub(
+            r"(// @\w+ )" + param + r"( \w+ )[^ ]+(.*)",  # noqa: W605
+            r"\g<1>" + param + r"\2" + regexps[param][2] + r"\3",
+            line)
+          replacements += 1
+          continue
+        prototype[i] = re.sub(regexps[param][0], regexps[param][1], line)
+        if line != prototype[i]:
+          replacements += 1
+    if replacements == 0:
+      raise Exception(
+          "No replacements made, are you sure you specified correct param?")
+    if replacements < len(regexps):
+      raise Warning("Made less replacements then number of params. Typo?")
+    temp_file = prototype_file + ".tmp"
+    with open(temp_file, "w") as w:
+      w.write("\n".join(prototype))
+    os.rename(temp_file, prototype_file)
+    logging.info("Successfully made %d replacements", replacements)
```

Comment:
What if replacement > len(regexps)? Could we accidentally replace something we're not supposed to?

Reply:
Got rid of this